nixos/modules/services/networking/nomad.nix

```diff
@@ -0,0 +1,74 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.nomad;
+  format = pkgs.formats.json { };
+in
+{
+  ##### interface
+  options = {
+    services.nomad = {
+      enable = mkEnableOption "Nomad, a distributed, highly available, datacenter-aware scheduler";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.nomad;
+        defaultText = "pkgs.nomad";
+        description = ''
+          The package used for the Nomad agent and CLI.
+        '';
+      };
+
+      dropPrivileges = mkOption {
+        type = types.bool;
+        default = true;
+        description = ''
+          Whether the nomad agent should be run as a non-root nomad user.
+        '';
+      };
+
+      settings = mkOption {
+        type = format.type;
+        default = { };
+        description = ''
+          Configuration for Nomad. See the <link xlink:href="https://www.nomadproject.io/docs/configuration">documentation</link>
+          for supported values.
+        '';
+      };
+    };
+  };
+
+  ##### implementation
+  config = mkIf cfg.enable {
+    environment = {
+      etc."nomad.json".source = format.generate "nomad.json" cfg.settings;
+      systemPackages = [ cfg.package ];
+    };
+
+    systemd.services.nomad = {
+      description = "Nomad";
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" ];
+      restartTriggers = [ environment.etc."nomad.json".source ];
+
+      serviceConfig = {
+        DynamicUser = cfg.dropPrivileges;
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        ExecStart = "${cfg.package}/bin/nomad agent -config=/etc/nomad.json";
+        KillMode = "process";
+        KillSignal = "SIGINT";
+        LimitNOFILE = 65536;
+        LimitNPROC = "infinity";
+        OOMScoreAdjust = -1000;
+        Restart = "on-failure";
+        RestartSec = 2;
+        StartLimitBurst = 3;
+        StartLimitIntervalSec = 10;
+        StateDirectory = "nomad";
```

Comment:
Even though this will materialize as `/var/lib/nomad`, nomad does not set `data_dir` by default so I think the unit should set `data_dir` in `cfg.settings` to `"/var/lib/nomad"`.

Reply:
Done.

A question though, @cpcloud:

Is what you do in https://github.com/cpcloud/nixpkgs/blob/f84c234a4f328b5c181cb9c955798cdfb5f57240/nixos/modules/services/networking/consul.nix#L7-L9 better:

```
  settings = {
    data_dir = "/var/lib/consul";
  } // cfg.settings;
in
      settings = mkOption {
        type = format.type;
        default = { };
```

or what I did with

```
      settings = mkOption {
        type = format.type;
        default = {
          data_dir = "/var/lib/nomad";
        };
```

? Does one have benefits vs the other? The `let`-bout one seems to have the drawback of not being obvious from documentation.

Reply:
That is a good question. I think that what you did is probably more obvious from a documentation perspective.

Reply:
You should probably set it this way:
```
config = mkIf cfg.enable {
  services.nomad.settings = {
    data_dir = "/var/lib/nomad";
  };
};
```

@nh2 if you do that then `data_dir` will be overridden if the user puts **anything** in `settings`. By setting it the way I have mentioned it gives the sysadmin an escape hatch where they can set `services.nomad.settings.data_dir = lib.mkForce "/something/else";` if they absolutely **need** to.